feat: dataset creation from traces

Juliaj opened the pull request:
## Purpose

- Address https://github.com/RobotecAI/rai/issues/663 for creating dataset for model fine-tuning.

## Proposed Changes

First attempt to extract observations from Langfuse traces and format data to create a dataset for model fine-tuning. A set of model fine-tuning related helper files are created to facilitate the verification of the dataset.

[README.md ](https://github.com/RobotecAI/rai/blob/3cd3539f2b5e1e8f7640dfa65698e6de494ed69a/src/rai_finetune/README.md)contains detailed documentation for these scripts.

@rachwalk, @maciejmajek, the PR is still a work in progress and I would like to get some early feedback from you specifically on how we should format the data and whether we should align data with specific areas (e.g. tool calling).

LLM fine-tuning is new to me, your feedback for the model tuning helpers are also very much appreciated. 

## Issues

- https://github.com/RobotecAI/rai/issues/663.

## Testing

- Traces generated from `gpt-4o-mini` with Manipulation-demo have been used to verify the extraction and formatting logic.
- qwen2.5:7b-instruct has been used to test the Python files under `/model`.

rachwalk commented on src/rai_finetune/README.md:
```diff
@@ -0,0 +1,209 @@
+
+## Environment Setup
+
+This module requires Python 3.11.9 (Python 3.13 does not support Unsloth). Use pyenv to manage Python versions:
```
Oficially RAI supports Python 3.10 (ROS2 humble) and Python 3.12 (ROS2 jazzy). This is due to issues with wheel compatibility issues. Have you conducted tests on these releases? I think it would be beneficial for consistency reasons to have `rai_finetune` use the same python version, or at least either of the two. Of course if that is possible -- if unsloth only works on 3.11.9 then not much we can do about it.

rachwalk replied:
Okay I've seen the comment in pyproject.toml and I see that 3.12 is not feasible, but 3.10 should work? Then If it's alright with you, for consistency's sake I'd keep this package officially supported for Python 3.10.

Juliaj replied:
Updated by [fc2e5a814b564e](https://github.com/RobotecAI/rai/pull/673/commits/fc2e5a814b564ecc0e3f145cce4d7942b88cd86c).

rachwalk commented on src/rai_finetune/rai_finetune/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# Copyright (C) 2025 RAI Development Team
+#
```
```suggestion
# Copyright (C) 2025 Julia Jia
#
```

While I truly appreciate this, unfortunately "RAI Development Team" does not posess the legal requirements to be a copyright holder. We've had a deep read into how Apache 2.0 works, and technically the Contributor is the Copyright Holder, unless it's explicitely transferred to another Contributor. So I make this suggestion, and if you're okay being listed as the holder then please update all notices in this PR.

If these files are modified in the future we will be using the following style (example if we modify it next year):


```
# Copyright (C) 2025 Julia Jia
# Copyright (C) 2026 Robotec.AI
#
```

Juliaj replied:
Thanks for the guidance, copyright headers have been updated.

rachwalk commented on src/rai_finetune/rai_finetune/data/langfuse_obs_extractor.py:
```diff
@@ -0,0 +1,489 @@
+    from langfuse import Langfuse
+except ImportError:
+    print("Error: langfuse package not found. Install with: pip install langfuse")
+    exit(1)
```
I am a bit confused, langfuse is listed as an obligatory dependency in pyproject.toml? Do you see it being an optional dependency in the future?

Juliaj replied:
Agreed, this is redundant, removed.

rachwalk commented on src/rai_finetune/rai_finetune/data/langfuse_obs_extractor.py:
```diff
@@ -0,0 +1,489 @@
+    # Observation related
+    output_file: str = "observations.jsonl"
+    max_observations: int = 1000
+    type_filter: str = "GENERATION"  # e.g., GENERATION, EVENT, SPAN
```
```suggestion
    type_filter: Literal["GENERATION", "EVENT", "SPAN" ] = "GENERATION"  # e.g., GENERATION, EVENT, SPAN
```

If I understand correctly there is a set list of valid options.

Juliaj replied:
Updated per your suggestion.

rachwalk commented on src/rai_finetune/rai_finetune/data/langfuse_obs_extractor.py:
```diff
@@ -0,0 +1,489 @@
+    max_observations: int = 1000
+    type_filter: str = "GENERATION"  # e.g., GENERATION, EVENT, SPAN
+    start_time: Optional[str] = None  # ISO format string
+    stop_time: Optional[str] = None  # ISO format string
```
Perhaps it would be more user friendly to use `datetime.datetime` or `datetime.time` and convert it to ISO string when required.

Juliaj replied:
By friendly, did you mean to make it easier for the user to input a short value ? I have this updated to take in both ISO string and `datetime.datetime`, see https://github.com/RobotecAI/rai/blob/8f5dbfb4dc9a08df4f6260041d2329bf1e657e4f/src/rai_finetune/rai_finetune/data/extractors/base.py#L98-L99 let me know what you think.

rachwalk commented on src/rai_finetune/rai_finetune/data/langfuse_obs_extractor.py:
```diff
@@ -0,0 +1,489 @@
+                "output",
+                "promptTokens",
+                "traceId",
+            ]
```
If the fields are not valid as None then please use default values, something like so. While the behaviour is the same, it helps with type hinting (e.g. Pyright won't complain about iterating over a None value).

```
from dataclasses import field
# ...
include_fields: list[str] =  field(default_factory=lambda: ["model", "input", "output", "promptTokens", "traceId"])
```

Juliaj replied:
Your point is taken. The fields are likely to be provider (Langfuse, Langsmith etc) specific, so for now I made this as an abstract method for subclass to override, see https://github.com/RobotecAI/rai/blob/8f5dbfb4dc9a08df4f6260041d2329bf1e657e4f/src/rai_finetune/rai_finetune/data/extractors/base.py#L66C1-L69C13. Wdyt?

rachwalk commented on src/rai_finetune/rai_finetune/data/langfuse_obs_extractor.py:
```diff
@@ -0,0 +1,489 @@
+def filter_observation_by_model(item: Any, filter_models: Optional[List[str]]) -> bool:
+    """Filter observations by model value present directly or inside metadata."""
+    if not filter_models:
+        return True
```
Is there a use case where one would want to call this function with `filter_observation_by_model(item, None)`? I am not sure if the filter list can be elsewhere automatically generated, such that it returns `None` -- in which case I understand the None check. However, if it's unlikely I'd opt for `filter_models: List[str]` as an obligatory argument.

Juliaj replied:
updated.

rachwalk commented on src/rai_finetune/rai_finetune/data/training_data_formatter.py:
```diff
@@ -0,0 +1,937 @@
+            # Handle dictionary content (e.g., function calls)
+            content = str(content)
+
+        return {"role": "tool", "content": str(content).strip()}
```
I think there are some portions of these functions which could be brought to a common functionality, to avoid repetition,

Juliaj replied:
Point taken. I have refactored this .py into a base class and a tool_calling formatter. If you could give them another look, it would be great.

rachwalk commented on src/rai_finetune/rai_finetune/data/training_data_formatter.py:
```diff
@@ -0,0 +1,937 @@
+                    )
+                else:
+                    # Skip unknown role types
+                    continue
```
Related to the comment above, I think a slightly more maintainable way of writing this would be to create a `self.formatted.format_message(content, self.config, role, **kwargs/tool_calls` -- though I am not dead set on it. Could you let me know why you've went with this design choice?

Juliaj replied:
I thought about this and introduced a new method to utllize chat template for formmating, hopefully this cuts down the complexity. https://github.com/RobotecAI/rai/blob/f16b6e81fd4b1b7e6468ec0b0346498e8ed078f2/src/rai_finetune/rai_finetune/data/formatters/base.py#L131-L147

rachwalk commented on src/rai_finetune/rai_finetune/data/training_data_formatter.py:
```diff
@@ -0,0 +1,937 @@
+class TrainingDataFormatter:
+    """Main class for formatting training data"""
+
+    def __init__(self, config: TrainingDataConfig):
```
I'd consider splitting this into 4 classes, each for the applicable format, instead of one class with a method for each format.

Juliaj replied:
See comments above, refactored the code.

rachwalk commented on src/rai_finetune/rai_finetune/model/ollama_model_producer.py:
```diff
@@ -0,0 +1,580 @@
+            }
+
+            with open(self.output_dir / "tools.json", "w") as f:
+                json.dump(tool_definitions, f, indent=2)
```
Correct me if I'm wrong, but this would only be relevant for use with one particular toolset. Wouldn't it be better for the user to specify a tools.json file? If you think that'd be valid, we could add a method do automatically generate such tools.json files from RAI Agents, making it easy to setup. Let me know what you think.

Juliaj replied:
Very valid points. I'm planning to refactor the code and to reduce the scope of this PR, I have the fine tune helper files removed. Will refactor and include them into subsequent PR.

rachwalk reviewed (changes requested):
Hi @Juliaj thank you for this PR. I think so far it's looking really good. I've left a few comments regarding stylistic changes. Overall I'd consider splitting some of the larger files containing multiple class definitions into separate ones -- for managability. It's a matter of preference, but I think it's better to work with smaller files, especially when large diffs come in place.

Another general comment I have would be creating abstract classes for functionalities you think could be well subclassed, i.e. Things related to specific models/providers (like langfuse may not be the only way to collect observations in the future, multiple formatters could be featured). That way we could already define an API for easy extensibility.

However to answer your question, I think that for this package and given RAI being an embodiment focused framework, I think that slight overfitting towards tool calling (and perhaps other embodying mechanisms such as the usage of `rai_whoami` in system prompts) would be beneficial.

Also thank you for your dilligent use of error reporting mechanisms and application of defensive programming, I think it's very cleanly implemented and we will be taking inspiration when it comes to development and maintance of `rai_core`.

Juliaj committed:
Refactored data extraction and formatting modules based on review feedback

Additional changes:
- Implemented a command-line interface in `data_cli.py` for data extraction and formatting.
- Updated README documentation and added a proposed folder structure

Juliaj committed:
Fix linting issue and remove old info from README

Juliaj committed:
Fix the smoke-test and add diagram to README

Juliaj committed:
Merge remote-tracking branch 'origin/main' into rai_dataset

Juliaj committed:
Fix the linting issue.

Juliaj committed:
Fix linting issue for import_test.py

rachwalk commented on src/rai_finetune/README.md:
```diff
@@ -0,0 +1,221 @@
+# RAI Fine-tuning Module
+
+## Module Overview
```
Please add in the header that this module is experimental and in development, and some features are in progress/might change working.

rachwalk commented on src/rai_finetune/pyproject.toml:
```diff
@@ -0,0 +1,62 @@
+
+[tool.poetry]
+name = "rai_finetune"
+version = "0.1.0"
```
I think it'd be better should be "0.1.0-alpha", as this is essentially a "prerelease" -- full fine tuning is not yet enabled.

rachwalk reviewed (approved):
I am approving the PR, however I left 2 comments regarding maintenance and documentation communiques. I'll be happy to merge after the comments are addressed.

maciejmajek commented:
+1, happy to merge after adding experimental info

Juliaj committed:
Update per review feedback

Juliaj commented:
> +1, happy to merge after adding experimental info

@maciejmajek, I've added an experimental info header to the README and updated the version to `0.1.0-alpha` as @rachwalk suggested. Please review when you have a chance.

maciejmajek committed:
Merge branch 'main' into rai_dataset
